examples/advanced/quickxplain.py

```diff
@@ -0,0 +1,83 @@
+"""
+    CPMpy implementation of the QuickXplain algorithm by Junker.
+        Junker, Ulrich. "Preferred explanations and relaxations for over-constrained problems." AAAI-2004. 2004.
+"""
+
+import cpmpy as cp
+from cpmpy.transformations.get_variables import get_variables
+from cpmpy.transformations.normalize import toplevel_list
+
+
+def quickXplain(soft, hard=[], order=lambda x : 0, solver="ortools"):
+    """
+        Find a preferred minimal unsatisfiable subset of constraints
+        A partial order is imposed on the constraints using the `order` argument
+        Constraints with lower order are preferred to ones with higher order
+
+        :param: soft: soft constraints, list of expressions
+        :param: hard: hard constraints, optional, list of expressions
+        :param: order: optional function to compute order of constraints
+        :param: solver: name of a solver, see SolverLookup.solvernames()
+            "z3" and "gurobi" are incremental, "ortools" restarts the solver
+    """
+
+    soft = toplevel_list(soft, merge_and=False)
+    assump = cp.boolvar(shape=len(soft))
+    solver = cp.SolverLookup.get(solver, cp.Model(hard))
+    solver += assump.implies(soft)
+
+    assert solver.solve(assumptions=assump) is False, "The model should be UNSAT!"
```

Comment:
here is a naive impelementation of the QuickXplain algo.

```Python
def quickXplain_naive(hard, soft, order=lambda x : 0, solver="ortools"):
    if len(soft) == 0: return []

    soft = toplevel_list(soft, merge_and=False)
    assert cp.Model(hard + soft).solve() is False, "The model should be UNSAT!"

    return recurse_explain_naive(hard, hard, soft, order)

def recurse_explain_naive(hard, delta, soft, order):

    if len(delta) != 0 and cp.Model(hard).solve() is False:
        # conflict is in hard constraints, no need to recurse
        return []

    if len(soft) == 1:
        # conflict is not in hard constraints, but only 1 soft constraint
        return list(soft) # base case of recursion

    soft = sorted(soft, key=lambda a : order(a))
    split = len(soft) // 2 # determine split point
    more_preffered, less_preffered = soft[:split], soft[split:] # split constraints into two sets

    # treat more preferred part as hard and find extra constants from less preferred
    delta2 = recurse_explain_naive(hard + more_preffered, more_preffered, less_preffered, order=order)
    # find which preferred constraints exactly
    delta1 = recurse_explain_naive(hard + delta2, delta2, more_preffered, order=order)
    return delta1 + delta2
```

Reply:
I think we should only add this in case the quickXplain algo becomes a "tool" in CPMpy, but not really needed for an example in my opinion.
Should we add it as a tool?

Reply:
Can be added to both examples and tools. I think it’s a rather compact example that would fit nicely in the tools.